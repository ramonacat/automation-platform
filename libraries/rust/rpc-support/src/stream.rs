use platform::async_infra::run_with_error_handling;
use thiserror::Error;
use tokio::io::{AsyncBufRead, AsyncBufReadExt, AsyncWrite, AsyncWriteExt};
use tokio::sync::mpsc::{Receiver, Sender};

use crate::wire_format::Frame;

#[derive(Error, Debug)]
pub enum Error {
    #[error("IO: {0:?}")]
    Io(#[from] std::io::Error),
    #[error("Serialization: {0:?}")]
    Serialization(#[from] serde_json::Error),
    #[error("frame send: {0:?}")]
    FrameSend(#[from] tokio::sync::mpsc::error::SendError<Frame>),
    #[error("join: {0:?}")]
    Join(#[from] tokio::task::JoinError),
}

pub(crate) struct Writer<TWriter: AsyncWrite> {
    writer: TWriter,
}

impl<TWriter: AsyncWrite + Unpin> Writer<TWriter> {
    pub fn from_raw(writer: TWriter) -> Self {
        Self { writer }
    }

    pub async fn write_frame(&mut self, frame: Frame) -> Result<(), Error> {
        let mut buffer = String::new();

        // TODO Replace unwraps with error handling
        buffer.push_str(&serde_json::to_string(&frame).unwrap());
        buffer.push('\n');

        self.writer.write_all(buffer.as_bytes()).await?;

        Ok(())
    }
}

pub struct Reader<TReader: AsyncBufRead> {
    reader: TReader,
}

impl<TReader: AsyncBufRead + Unpin> Reader<TReader> {
    pub fn from_raw(reader: TReader) -> Self {
        Self { reader }
    }

    /// # Errors
    /// Can fail if the connection is dropped
    pub async fn read_frame(&mut self) -> Result<Frame, Error> {
        let mut line_buffer = String::new();

        self.reader.read_line(&mut line_buffer).await?;
        let frame = serde_json::from_str(&line_buffer)?;

        Ok(frame)
    }
}

pub struct Stream<TRawReader: AsyncBufRead, TRawWriter: AsyncWrite> {
    reader: Reader<TRawReader>,
    writer: Writer<TRawWriter>,
}

impl<
        TRawReader: AsyncBufRead + Unpin + Send + 'static,
        TRawWriter: AsyncWrite + Unpin + Send + 'static,
    > Stream<TRawReader, TRawWriter>
{
    pub fn from_reader_writer(raw_reader: TRawReader, raw_writer: TRawWriter) -> Self {
        Self {
            reader: Reader::from_raw(raw_reader),
            writer: Writer::from_raw(raw_writer),
        }
    }

    async fn reader_task(
        mut reader: Reader<TRawReader>,
        read_tx: Sender<Frame>,
    ) -> Result<(), Error> {
        while let Ok(frame) = reader.read_frame().await {
            read_tx.send(frame).await?;
        }

        Ok(())
    }

    async fn writer_task(
        mut writer: Writer<TRawWriter>,
        mut write_rx: Receiver<Frame>,
    ) -> Result<(), Error> {
        while let Some(frame) = write_rx.recv().await {
            writer.write_frame(frame).await?;
        }

        Ok(())
    }

    /// # Errors
    /// Can fail if the background tasks fail
    pub async fn run(self, read_tx: Sender<Frame>, write_rx: Receiver<Frame>) -> Result<(), Error> {
        let Self { reader, writer } = self;

        let reader_task = tokio::spawn(run_with_error_handling(Self::reader_task(reader, read_tx)));
        let writer_task =
            tokio::spawn(run_with_error_handling(Self::writer_task(writer, write_rx)));

        let join_result = tokio::join!(reader_task, writer_task);

        join_result.0?;
        join_result.1?;

        Ok(())
    }
}
