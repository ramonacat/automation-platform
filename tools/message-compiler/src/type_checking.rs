use crate::parsing::{
    EnumDefinitionRaw, EnumVariantRaw, FieldRaw, FileRaw, IdentifierRaw, StructDefinitionRaw,
    TypeRaw,
};
use petgraph::algo::toposort;
use petgraph::graph::DiGraph;
use std::collections::HashMap;
use std::error::Error;
use std::fmt::{Display, Formatter};

#[derive(Debug)]
pub enum TypeCheckError {
    RepeatedName(String),
    RepeatedFieldName {
        field_name: String,
        struct_name: String,
    },
    StructNotFound(String),
}

impl Display for TypeCheckError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TypeCheckError::RepeatedName(name) => {
                write!(f, "The type with name \"{name}\" already exists")
            }
            TypeCheckError::RepeatedFieldName {
                field_name,
                struct_name,
            } => write!(
                f,
                "A field with name \"{field_name}\" already exists in struct \"{struct_name}\""
            ),
            TypeCheckError::StructNotFound(name) => {
                write!(f, "A struct with name \"{name}\" does not exist")
            }
        }
    }
}

impl Error for TypeCheckError {}

#[derive(Debug)]
pub enum TypedFieldType {
    U8,
    U16,
    U32,
    U64,
    S8,
    S16,
    S32,
    S64,
    Instant,
    Guid,
    String,
    Void,
    Binary,
    OtherStruct(String),
    Enum(String),
    Optional(Box<TypedFieldType>),
    Array(Box<TypedFieldType>),
}

#[derive(Debug)]
pub struct TypedField {
    pub name: String,
    pub type_id: TypedFieldType,
}

impl TypedField {
    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    #[must_use]
    pub fn type_name(&self) -> &TypedFieldType {
        &self.type_id
    }
}

#[derive(Debug)]
pub struct TypedStruct {
    pub name: String,
    pub fields: Vec<TypedField>,
}

impl TypedStruct {
    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    #[must_use]
    pub fn fields(&self) -> &[TypedField] {
        &self.fields
    }
}

#[derive(Debug)]
pub struct TypedEnumVariant {
    pub name: String,
    pub fields: Vec<TypedField>,
}

impl TypedEnumVariant {
    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    #[must_use]
    pub fn fields(&self) -> &[TypedField] {
        &self.fields
    }
}

#[derive(Debug)]
pub struct TypedEnum {
    pub name: String,
    pub variants: Vec<TypedEnumVariant>,
}

impl TypedEnum {
    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    #[must_use]
    pub fn variants(&self) -> &[TypedEnumVariant] {
        &self.variants
    }
}

#[derive(Debug)]
enum TypeCheckableFieldType<'a> {
    U8,
    U16,
    U32,
    U64,
    S8,
    S16,
    S32,
    S64,
    Instant,
    Guid,
    String,
    Void,
    Binary,
    ToBeResolved(&'a str),
    Optional(Box<TypeCheckableFieldType<'a>>),
    Array(Box<TypeCheckableFieldType<'a>>),
}

#[derive(Debug)]
struct TypeCheckableStructDefinition<'input> {
    name: String,
    fields: HashMap<&'input str, TypeCheckableFieldType<'input>>,
}

#[derive(Debug)]
struct TypeCheckableEnumVariant<'input> {
    name: String,
    fields: HashMap<&'input str, TypeCheckableFieldType<'input>>,
}

#[derive(Debug)]
struct TypeCheckableEnumDefinition<'input> {
    name: String,
    variants: HashMap<&'input str, TypeCheckableEnumVariant<'input>>,
}

#[derive(Debug)]
enum TypeCheckableDataDefinition<'input> {
    Struct(&'input TypeCheckableStructDefinition<'input>),
    Enum(&'input TypeCheckableEnumDefinition<'input>),
}

pub struct TypedMetadata {
    pub fields: Vec<TypedField>,
}

impl TypedMetadata {
    #[must_use]
    pub fn fields(&self) -> &[TypedField] {
        &self.fields
    }
}

pub enum TypedRpcCall {
    Stream {
        name: String,
        request: TypedFieldType,
        response: TypedFieldType,
    },
    Unary {
        name: String,
        request: TypedFieldType,
        response: TypedFieldType,
    },
}

pub struct TypedRpc {
    pub calls: Vec<TypedRpcCall>,
}

impl TypedRpc {
    #[must_use]
    pub fn calls(&self) -> &[TypedRpcCall] {
        &self.calls
    }
}

pub struct TypedFile {
    pub structs: Vec<TypedStruct>,
    pub enums: Vec<TypedEnum>,
    pub meta: TypedMetadata,
    pub rpc: TypedRpc,
}

#[derive(Default)]
pub struct TypeChecker<'input> {
    structs: HashMap<String, TypeCheckableStructDefinition<'input>>,
    enums: HashMap<String, TypeCheckableEnumDefinition<'input>>,
}

impl<'input> TypeChecker<'input> {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    fn check_duplicate(&'input self, name: &'input IdentifierRaw) -> Result<(), TypeCheckError> {
        if self.structs.contains_key(name.0) {
            return Err(TypeCheckError::RepeatedName(name.0.to_string()));
        }

        Ok(())
    }

    fn map_fields(
        fields_raw: &[FieldRaw<'input>],
        struct_name: &str,
    ) -> Result<HashMap<&'input str, TypeCheckableFieldType<'input>>, TypeCheckError> {
        let mut fields = HashMap::new();

        for field_raw in fields_raw {
            let type_id = Self::resolve_raw_type(&field_raw.type_name);

            if fields.contains_key(field_raw.name.0) {
                return Err(TypeCheckError::RepeatedFieldName {
                    field_name: field_raw.name.0.to_string(),
                    struct_name: struct_name.to_string(),
                });
            }
            fields.insert(field_raw.name.0, type_id);
        }

        Ok(fields)
    }

    fn raw_field_type_to_typecheckable_type(
        type_name: &IdentifierRaw<'input>,
    ) -> TypeCheckableFieldType<'input> {
        match type_name.0 {
            "u8" => TypeCheckableFieldType::U8,
            "u16" => TypeCheckableFieldType::U16,
            "u32" => TypeCheckableFieldType::U32,
            "u64" => TypeCheckableFieldType::U64,
            "s8" => TypeCheckableFieldType::S8,
            "s16" => TypeCheckableFieldType::S16,
            "s32" => TypeCheckableFieldType::S32,
            "s64" => TypeCheckableFieldType::S64,
            "instant" => TypeCheckableFieldType::Instant,
            "guid" => TypeCheckableFieldType::Guid,
            "string" => TypeCheckableFieldType::String,
            "void" => TypeCheckableFieldType::Void,
            "binary" => TypeCheckableFieldType::Binary,
            other => TypeCheckableFieldType::ToBeResolved(other),
        }
    }

    fn resolve_raw_type(type_raw: &TypeRaw<'input>) -> TypeCheckableFieldType<'input> {
        match type_raw {
            TypeRaw::Type(id) => Self::raw_field_type_to_typecheckable_type(id),
            TypeRaw::Optional(id) => TypeCheckableFieldType::Optional(Box::new(
                Self::raw_field_type_to_typecheckable_type(id),
            )),
            TypeRaw::Array(id) => TypeCheckableFieldType::Array(Box::new(
                Self::raw_field_type_to_typecheckable_type(id),
            )),
        }
    }

    fn type_check_fields(
        &self,
        raw_fields: &HashMap<&str, TypeCheckableFieldType>,
    ) -> Result<Vec<TypedField>, TypeCheckError> {
        let mut fields = vec![];

        for (field_name, field_type) in raw_fields {
            fields.push(TypedField {
                name: (*field_name).to_string(),
                type_id: self.resolve_type(field_type)?,
            });
        }

        Ok(fields)
    }

    fn resolve_type(
        &self,
        field_type: &TypeCheckableFieldType,
    ) -> Result<TypedFieldType, TypeCheckError> {
        Ok(match field_type {
            TypeCheckableFieldType::U8 => TypedFieldType::U8,
            TypeCheckableFieldType::U16 => TypedFieldType::U16,
            TypeCheckableFieldType::U32 => TypedFieldType::U32,
            TypeCheckableFieldType::U64 => TypedFieldType::U64,
            TypeCheckableFieldType::S8 => TypedFieldType::S8,
            TypeCheckableFieldType::S16 => TypedFieldType::S16,
            TypeCheckableFieldType::S32 => TypedFieldType::S32,
            TypeCheckableFieldType::S64 => TypedFieldType::S64,
            TypeCheckableFieldType::Instant => TypedFieldType::Instant,
            TypeCheckableFieldType::Guid => TypedFieldType::Guid,
            TypeCheckableFieldType::String => TypedFieldType::String,
            TypeCheckableFieldType::Void => TypedFieldType::Void,
            TypeCheckableFieldType::Binary => TypedFieldType::Binary,
            TypeCheckableFieldType::ToBeResolved(type_name) => {
                if self.structs.contains_key(*type_name) {
                    return Ok(TypedFieldType::OtherStruct((*type_name).to_string()));
                } else if self.enums.contains_key(*type_name) {
                    return Ok(TypedFieldType::Enum((*type_name).to_string()));
                }
                return Err(TypeCheckError::StructNotFound((*type_name).to_string()));
            }
            TypeCheckableFieldType::Optional(type_) => {
                let type_id = self.resolve_type(type_)?;
                TypedFieldType::Optional(Box::new(type_id))
            }
            TypeCheckableFieldType::Array(type_) => {
                let type_id = self.resolve_type(type_)?;
                TypedFieldType::Array(Box::new(type_id))
            }
        })
    }

    /// # Errors
    /// May return an error when the type check fails
    /// # Panics
    /// TODO MAKE THIS NOT EVER PANIC
    /// todo split into smaller functions
    #[allow(clippy::too_many_lines)]
    pub fn check(mut self, file: &FileRaw<'input>) -> Result<TypedFile, TypeCheckError> {
        for StructDefinitionRaw(name, fields) in file.structs() {
            self.check_duplicate(name)?;

            let fields = Self::map_fields(fields, name.0)?;

            self.structs.insert(
                name.0.to_string(),
                TypeCheckableStructDefinition {
                    name: name.0.to_string(),
                    fields,
                },
            );
        }

        for EnumDefinitionRaw { name, variants } in file.enums() {
            // fixme check for name conflicts with structs
            // fixme check for name conflicts with other enums

            let variants = Self::map_enum_variants(variants, name.0)?;

            self.enums.insert(
                name.0.to_string(),
                TypeCheckableEnumDefinition {
                    name: name.0.to_string(),
                    variants,
                },
            );
        }

        let mut metadata_fields = HashMap::new();
        if let Some(metadata) = file.metadata() {
            metadata_fields = Self::map_fields(metadata.fields(), "metadata")?;
        }

        let mut graph = DiGraph::new();
        let mut node_ids = HashMap::new();
        for struct_definition in self.structs.values() {
            let ix = graph.add_node(TypeCheckableDataDefinition::Struct(struct_definition));
            node_ids.insert(struct_definition.name.to_string(), ix);
        }

        for enum_definition in self.enums.values() {
            let ix = graph.add_node(TypeCheckableDataDefinition::Enum(enum_definition));
            node_ids.insert(enum_definition.name.to_string(), ix);
        }

        for struct_definition in self.structs.values() {
            for field_definition in struct_definition.fields.values() {
                if let TypeCheckableFieldType::ToBeResolved(name) = field_definition {
                    graph.add_edge(
                        *node_ids.get(*name).unwrap(),
                        *node_ids.get(&struct_definition.name).unwrap(),
                        (*name).to_string(),
                    );
                }
            }
        }

        let sorted = toposort(&graph, None).unwrap();
        let mut structs_typed = HashMap::new();
        let mut enums_typed = HashMap::new();

        for x in sorted {
            let node = graph.node_weight(x).unwrap();

            match node {
                TypeCheckableDataDefinition::Struct(struct_node) => {
                    let typed_struct = TypedStruct {
                        name: struct_node.name.clone(),
                        fields: self.type_check_fields(&struct_node.fields)?,
                    };
                    structs_typed.insert(struct_node.name.clone(), typed_struct);
                }
                TypeCheckableDataDefinition::Enum(enum_node) => {
                    let variants = enum_node.variants.iter().map(|variant| {
                        let fields = self.type_check_fields(&variant.1.fields).unwrap();

                        TypedEnumVariant {
                            name: variant.1.name.to_string(),
                            fields,
                        }
                    });

                    enums_typed.insert(
                        enum_node.name.clone(),
                        TypedEnum {
                            name: enum_node.name.clone(),
                            variants: variants.collect(),
                        },
                    );
                }
            }
        }

        let meta_fields = self.type_check_fields(&metadata_fields)?;
        let rpc = file.rpc().map(|rpc| {
            let mut rpc_typed = HashMap::new();
            for rpc_definition in &rpc.definitions {
                match rpc_definition {
                    crate::parsing::RpcDefinitionRaw::Stream {
                        name,
                        request,
                        response,
                    } => {
                        let typed_rpc = TypedRpcCall::Stream {
                            name: name.0.to_string(),
                            request: self.resolve_type(&Self::resolve_raw_type(request)).unwrap(),
                            response: self
                                .resolve_type(&Self::resolve_raw_type(response))
                                .unwrap(),
                        };
                        rpc_typed.insert(name.0.to_string(), typed_rpc);
                    }
                    crate::parsing::RpcDefinitionRaw::Unary {
                        name,
                        request,
                        response,
                    } => {
                        let typed_rpc = TypedRpcCall::Unary {
                            name: name.0.to_string(),
                            request: self.resolve_type(&Self::resolve_raw_type(request)).unwrap(),
                            response: self
                                .resolve_type(&Self::resolve_raw_type(response))
                                .unwrap(),
                        };
                        rpc_typed.insert(name.0.to_string(), typed_rpc);
                    }
                };
            }

            rpc_typed
        });

        Ok(TypedFile {
            structs: structs_typed.into_values().collect(),
            enums: enums_typed.into_values().collect(),
            meta: TypedMetadata {
                fields: meta_fields,
            },
            rpc: TypedRpc {
                calls: rpc.map_or_else(Vec::new, |rpc| rpc.into_values().collect()),
            },
        })
    }

    fn map_enum_variants(
        variants: &[EnumVariantRaw<'input>],
        name: &str,
    ) -> Result<HashMap<&'input str, TypeCheckableEnumVariant<'input>>, TypeCheckError> {
        variants
            .iter()
            .map(|variant| {
                let fields = Self::map_fields(&variant.fields, name)?;
                Ok((
                    variant.name.0,
                    TypeCheckableEnumVariant {
                        name: variant.name.0.to_string(),
                        fields,
                    },
                ))
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn type_check_error_display_tests() {
        let error = TypeCheckError::RepeatedName("test".to_string());

        assert_eq!(
            error.to_string(),
            "The type with name \"test\" already exists"
        );

        let error = TypeCheckError::RepeatedFieldName {
            field_name: "test1".to_string(),
            struct_name: "test2".to_string(),
        };

        assert_eq!(
            error.to_string(),
            "A field with name \"test1\" already exists in struct \"test2\""
        );
    }
}
