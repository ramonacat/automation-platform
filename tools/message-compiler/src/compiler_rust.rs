use proc_macro2::TokenStream;
use quote::{format_ident, quote, TokenStreamExt};

use crate::type_checking::{
    TypedEnum, TypedFieldType, TypedFile, TypedMetadata, TypedRpc, TypedStruct,
};

fn generate_header() -> TokenStream {
    quote! {
        #[allow(unused)]
        use async_std::stream::Stream;
        use rpc_support::rpc_error::RpcError;
        use serde::{Deserialize, Serialize};
    }
}

fn generate_metadata(meta: &TypedMetadata) -> TokenStream {
    let mut result = quote!(
        #[derive(Serialize, Deserialize, Debug, Clone)]
    );

    let mut meta_fields = quote! {};
    for f in meta.fields() {
        let name = format_ident!("{}", f.name());
        let ty: syn::Type = syn::parse_str(&to_rust_type(f.type_name())).unwrap();
        meta_fields.append_all(quote!(pub #name: #ty,));
    }
    result.append_all(quote!(
        pub struct Metadata {
            #meta_fields
        }
    ));

    result
}

fn generate_structs(structs: &Vec<TypedStruct>) -> TokenStream {
    let mut result = quote! {};

    for s in structs {
        let struct_name = format_ident!("{}", s.name());
        let mut render_fields = quote! {};
        for f in s.fields() {
            let name = format_ident!("{}", f.name());
            let ty: syn::Type = syn::parse_str(&to_rust_type(f.type_name())).unwrap();
            render_fields.append_all(quote!(pub #name: #ty,));
        }
        result.append_all(quote!(
            #[derive(Serialize, Deserialize, Debug, Clone)]
            pub struct #struct_name {
                #render_fields
            }
        ));
    }

    result
}

/// # Panics
/// TODO make this not panic
#[must_use]
pub fn generate_enums(enums: &Vec<TypedEnum>) -> TokenStream {
    let mut result = quote! {};

    for e in enums {
        let enum_name = format_ident!("{}", e.name());
        let mut render_variants = quote! {};
        for v in e.variants() {
            let variant_name = format_ident!("{}", v.name());
            let mut render_fields = quote! {};
            for f in v.fields() {
                let name = format_ident!("{}", f.name());
                let ty: syn::Type = syn::parse_str(&to_rust_type(f.type_name())).unwrap();
                render_fields.append_all(quote!(#name: #ty,));
            }
            render_variants.append_all(quote!(
                #variant_name {
                    #render_fields
                },
            ));
        }
        result.append_all(quote!(
            #[derive(Serialize, Deserialize, Debug, Clone)]
            pub enum #enum_name {
                #render_variants
            }
        ));
    }

    result
}

fn generate_rpc_trait(rpc: &TypedRpc) -> TokenStream {
    let mut rpc_methods = quote! {};

    for r in rpc.calls() {
        match r {
            crate::type_checking::TypedRpcCall::Stream {
                name,
                request,
                response,
            } => {
                let name = format_ident!("{}", name);
                let request: syn::Type = syn::parse_str(&to_rust_type(request)).unwrap();
                let response: syn::Type = syn::parse_str(&to_rust_type(response)).unwrap();

                rpc_methods.append_all(quote!(
                    async fn #name(
                        &mut self,
                        request: #request,
                        metadata: Metadata,
                    ) -> Result<
                        std::pin::Pin<Box<dyn Stream<Item = Result<#response, RpcError>> + Unpin + Send>>,
                        RpcError,
                    >;
                ));
            }
            crate::type_checking::TypedRpcCall::Unary {
                name,
                request,
                response,
            } => {
                let name = format_ident!("{}", name);
                let request: syn::Type = syn::parse_str(&to_rust_type(request)).unwrap();
                let response: syn::Type = syn::parse_str(&to_rust_type(response)).unwrap();

                rpc_methods.append_all(quote!(
                    async fn #name(
                        &mut self,
                        request: #request,
                        metadata: Metadata,
                    ) -> Result<#response, RpcError>;
                ));
            }
        }
    }

    quote!(
        #[async_trait::async_trait]
        pub trait Rpc {
             #rpc_methods
        }
    )
}

fn generate_rpc_client(rpc: &TypedRpc) -> TokenStream {
    let mut result = quote! {
        #[allow(unused)] use rpc_support::RawRpcClient;
        #[allow(unused)] use std::sync::atomic::{AtomicU64, Ordering};
        #[allow(unused)] use std::pin::Pin;

        pub struct Client<TRpcClient> where TRpcClient: RawRpcClient + Send + Sync {
            id: AtomicU64,
            raw: TRpcClient,
        }
    };

    result.append_all(quote! {
        impl<TRpcClient> Client<TRpcClient> where TRpcClient: RawRpcClient + Send + Sync {
            pub fn new(raw: TRpcClient) -> Self {
                Self {
                    id: AtomicU64::new(0),
                    raw,
                }
            }
        }
    });

    let mut rpc_methods = quote!();

    for method in &rpc.calls {
        let method_impl = match method {
            crate::type_checking::TypedRpcCall::Stream {
                name,
                request,
                response,
            } => {
                let name_ident = format_ident!("{}", name);
                let request: syn::Type = syn::parse_str(&to_rust_type(request)).unwrap();
                let response: syn::Type = syn::parse_str(&to_rust_type(response)).unwrap();

                quote! {
                    async fn #name_ident(
                        &mut self,
                        request: #request,
                        metadata: Metadata,
                    ) -> Result<Pin<Box<dyn Stream<Item = Result<#response, RpcError>> + Unpin + Send>>, RpcError> {
                        self.raw
                            .send_rpc_stream_request(
                                self.id.fetch_add(1, Ordering::AcqRel),
                                #name,
                                &request,
                                &metadata,
                            )
                            .await
                    }
                }
            }
            crate::type_checking::TypedRpcCall::Unary {
                name,
                request,
                response,
            } => {
                let name_ident = format_ident!("{}", name);
                let request: syn::Type = syn::parse_str(&to_rust_type(request)).unwrap();
                let response: syn::Type = syn::parse_str(&to_rust_type(response)).unwrap();

                quote! {
                    async fn #name_ident(&mut self, request: #request, metadata: Metadata) -> Result<#response, RpcError> {
                        self.raw
                            .send_rpc(
                                self.id.fetch_add(1, Ordering::AcqRel),
                                #name,
                                &request,
                                &metadata,
                            )
                            .await
                    }
                }
            }
        };

        rpc_methods.append_all(method_impl);
    }

    result.append_all(quote! {
        #[async_trait::async_trait]
        impl<TRpcClient> Rpc for Client<TRpcClient> where TRpcClient: RawRpcClient + Send + Sync {
            #rpc_methods
        }
    });

    result
}

fn generate_rpc_server_method_match(rpc: &TypedRpc) -> TokenStream {
    let mut method_cases = quote! {};

    for method in &rpc.calls {
        let case = match method {
            crate::type_checking::TypedRpcCall::Stream {
                name,
                request: _,
                response: _,
            } => {
                let name_ident = format_ident!("{}", name);

                quote! {
                    #name => {
                        let result = rpc.lock().await.#name_ident(serde_json::from_str(&payload_line)?, metadata).await;

                        send_stream_response(&mut write, result, request_id).await?;
                    }
                }
            }
            crate::type_checking::TypedRpcCall::Unary {
                name,
                request: _,
                response: _,
            } => {
                let name_ident = format_ident!("{}", name);

                quote! {
                    #name => {
                        let result = rpc.lock().await.#name_ident(serde_json::from_str(&payload_line)?, metadata).await;

                        send_response(&mut write, result, request_id, false).await?;
                    }
                }
            }
        };
        method_cases.append_all(case);
    }

    quote! {
        match method_name.as_str() {
            #method_cases

            // fixme do not panic here!
            _ => panic!("Unknown method name: {method_name}"),
        }
    }
}

fn generate_rpc_server(rpc: &TypedRpc) -> TokenStream {
    let method_match = generate_rpc_server_method_match(rpc);

    let result = quote! {
        use tracing::info;
        use thiserror::Error;
        use std::sync::Arc;
        use tokio::sync::Mutex;
        use tokio::net::TcpListener;
        use tokio::net::TcpStream;
        use tokio::io::BufReader;
        use rpc_support::send_response;
        #[allow(unused)] use rpc_support::send_stream_response;
        use rpc_support::read_request;

        pub struct Server<TRpc>
        where
            TRpc: Rpc + Send + Sync,
        {
            tcp: Arc<Mutex<TcpListener>>,
            rpc: Arc<Mutex<TRpc>>,
        }

        #[derive(Debug, Error)]
        pub enum ClientError {
            #[error("{0}")]
            IoError(#[from] tokio::io::Error),
            #[error("{0}")]
            JsonError(#[from] serde_json::Error),
            #[error("{0}")]
            RpcError(#[from] RpcError),
        }

        #[derive(Debug, Error)]
        pub enum RunError {
            #[error("{0}")]
            IoError(#[from] tokio::io::Error),
        }

        impl<T> Server<T>
        where
            T: Rpc + Send + Sync + 'static,
        {
            /// # Errors
            /// Will return an error when establishing the TCP Listener fails
            pub async fn new(addr: &str, rpc: Arc<Mutex<T>>) -> Result<Self, RpcError> {
                let tcp = Arc::new(Mutex::new(TcpListener::bind(addr).await?));
                Ok(Server { tcp, rpc })
            }

            async fn handle_client(socket: TcpStream, rpc: Arc<Mutex<T>>) -> Result<(), ClientError> {
                let mut socket = socket;
                let (read, mut write) = socket.split();
                let mut reader = BufReader::new(read);

                loop {
                    let (payload_line, method_name, request_id, metadata): (String, String, u64, Metadata) =
                        read_request(&mut reader).await?;

                    #method_match
                }
            }

            /// # Errors
            /// Will return an error if the connection fails
            pub async fn run(self) -> Result<(), RunError> {
                loop {
                    let (socket, address) = self.tcp.lock().await.accept().await?;
                    info!("New client connected: {}", address);

                    let rpc = self.rpc.clone();

                    tokio::spawn(platform::async_infra::run_with_error_handling(
                        Self::handle_client(socket, rpc),
                    ));
                }
            }
        }
    };

    result
}

#[must_use]
/// # Panics
/// TODO make this not panic
pub fn compile(file: TypedFile) -> String {
    let TypedFile {
        structs,
        meta,
        rpc,
        enums,
    } = file;

    let mut result = generate_header();

    result.append_all(generate_metadata(&meta));
    result.append_all(generate_structs(&structs));
    result.append_all(generate_enums(&enums));
    result.append_all(generate_rpc_trait(&rpc));

    result.append_all(generate_rpc_client(&rpc));
    result.append_all(generate_rpc_server(&rpc));

    prettyplease::unparse(&syn::parse_file(&result.to_string()).unwrap())
}

fn to_rust_type(type_: &TypedFieldType) -> String {
    match type_ {
        TypedFieldType::U8 => "u8".to_string(),
        TypedFieldType::U16 => "u16".to_string(),
        TypedFieldType::U32 => "u32".to_string(),
        TypedFieldType::U64 => "u64".to_string(),
        TypedFieldType::S8 => "s8".to_string(),
        TypedFieldType::S16 => "s16".to_string(),
        TypedFieldType::S32 => "s32".to_string(),
        TypedFieldType::S64 => "s64".to_string(),
        TypedFieldType::Instant => "std::time::SystemTime".to_string(),
        TypedFieldType::Guid => "uuid::Uuid".to_string(),
        TypedFieldType::String => "String".to_string(),
        TypedFieldType::Void => "()".to_string(),
        TypedFieldType::Binary => "Vec<u8>".to_string(),
        TypedFieldType::OtherStruct(name) | TypedFieldType::Enum(name) => name.clone(),
        TypedFieldType::Optional(type_) => format!("Option<{}>", to_rust_type(type_)),
        TypedFieldType::Array(type_) => format!("Vec<{}>", to_rust_type(type_)),
    }
}

#[cfg(test)]
mod tests {
    use crate::type_checking::TypedField;

    use super::*;

    #[test]
    pub fn to_rust_type_tests() {
        assert_eq!(to_rust_type(&TypedFieldType::U8), "u8");
        assert_eq!(to_rust_type(&TypedFieldType::U16), "u16");
        assert_eq!(to_rust_type(&TypedFieldType::U32), "u32");
        assert_eq!(to_rust_type(&TypedFieldType::U64), "u64");
        assert_eq!(to_rust_type(&TypedFieldType::S8), "s8");
        assert_eq!(to_rust_type(&TypedFieldType::S16), "s16");
        assert_eq!(to_rust_type(&TypedFieldType::S32), "s32");
        assert_eq!(to_rust_type(&TypedFieldType::S64), "s64");
        assert_eq!(
            to_rust_type(&TypedFieldType::Instant),
            "std::time::SystemTime"
        );
        assert_eq!(to_rust_type(&TypedFieldType::Guid), "uuid::Uuid");
        assert_eq!(to_rust_type(&TypedFieldType::String), "String");
        assert_eq!(to_rust_type(&TypedFieldType::Void), "()");
        assert_eq!(to_rust_type(&TypedFieldType::Binary), "Vec<u8>");
        assert_eq!(
            to_rust_type(&TypedFieldType::OtherStruct("Foo".to_string())),
            "Foo"
        );
        assert_eq!(
            to_rust_type(&TypedFieldType::Enum("Foo".to_string())),
            "Foo"
        );
        assert_eq!(
            to_rust_type(&TypedFieldType::Optional(Box::new(TypedFieldType::U8))),
            "Option<u8>"
        );
        assert_eq!(
            to_rust_type(&TypedFieldType::Array(Box::new(TypedFieldType::U8))),
            "Vec<u8>"
        );
    }

    #[test]
    pub fn generate_header_test() {
        insta::assert_snapshot!(prettyplease::unparse(
            &syn::parse_file(&generate_header().to_string()).unwrap()
        ));
    }

    #[test]
    pub fn generate_metadata_test() {
        let meta = generate_metadata(&TypedMetadata {
            fields: vec![
                TypedField {
                    name: "foo".to_string(),
                    type_id: TypedFieldType::U8,
                },
                TypedField {
                    name: "bar".to_string(),
                    type_id: TypedFieldType::U64,
                },
            ],
        });

        insta::assert_snapshot!(prettyplease::unparse(
            &syn::parse_file(&meta.to_string()).unwrap()
        ));
    }

    #[test]
    pub fn generate_structs_test() {
        let structs = generate_structs(&vec![
            TypedStruct {
                name: "Foo".to_string(),
                fields: vec![
                    TypedField {
                        name: "foo".to_string(),
                        type_id: TypedFieldType::U8,
                    },
                    TypedField {
                        name: "bar".to_string(),
                        type_id: TypedFieldType::U64,
                    },
                ],
            },
            TypedStruct {
                name: "Bar".to_string(),
                fields: vec![
                    TypedField {
                        name: "foo".to_string(),
                        type_id: TypedFieldType::U8,
                    },
                    TypedField {
                        name: "bar".to_string(),
                        type_id: TypedFieldType::U64,
                    },
                ],
            },
        ]);

        insta::assert_snapshot!(prettyplease::unparse(
            &syn::parse_file(&structs.to_string()).unwrap()
        ));
    }
}
