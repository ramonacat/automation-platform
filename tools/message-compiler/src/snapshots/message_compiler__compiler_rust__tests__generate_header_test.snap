---
source: src/compiler_rust.rs
expression: "prettyplease::unparse(&syn::parse_file(&generate_header().to_string()).unwrap())"
---
#[allow(unused)]
use async_std::stream::Stream;
use rpc_support::rpc_error::RpcError;
use serde::{Deserialize, Serialize};

